flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/writer/RecordWriter.java

```diff
@@ -111,93 +117,49 @@ public void randomEmit(T record) throws IOException, InterruptedException {
 	private void sendToTarget(T record, int targetChannel) throws IOException, InterruptedException {
 		RecordSerializer<T> serializer = serializers[targetChannel];
 
-		synchronized (serializer) {
-			SerializationResult result = serializer.addRecord(record);
-
-			while (result.isFullBuffer()) {
-				Buffer buffer = serializer.getCurrentBuffer();
-
-				if (buffer != null) {
-					numBytesOut.inc(buffer.getSizeUnsafe());
-					writeAndClearBuffer(buffer, targetChannel, serializer);
-
-					// If this was a full record, we are done. Not breaking
-					// out of the loop at this point will lead to another
-					// buffer request before breaking out (that would not be
-					// a problem per se, but it can lead to stalls in the
-					// pipeline).
-					if (result.isFullRecord()) {
-						break;
-					}
-				} else {
-					BufferBuilder bufferBuilder =
-						targetPartition.getBufferProvider().requestBufferBuilderBlocking();
-					result = serializer.setNextBufferBuilder(bufferBuilder);
+		SerializationResult result = serializer.addRecord(record);
+
+		while (result.isFullBuffer()) {
```

Comment:
As we discussed, I'm not entirely sure. This "minor change" can be a significant overhead in case of many channels and large records. I don't want to risk increasing the scope of potential problems with this PR :(

Reply:
👍 Can introduce this change later after some more extensive tests.